overview and quickstart (#155)
leaningtech/labs: sites/cheerpx/src/content/docs/00-overview.mdx

---

import LinkButton from "@leaningtech/astro-theme/components/LinkButton.astro";
import { DISCORD_URL } from "@/consts.ts";

<div class="not-prose flex gap-2 mb-2">
	<img src="https://img.shields.io/badge/version-1.0-blue" alt="Version 1.0" />
	<a href="https://discord.leaningtech.com/">
		<img
			src="https://img.shields.io/discord/988743885121548329?color=%237289DA&logo=discord&logoColor=ffffff"
			alt="Discord server"
		/>
	</a>
	<a href="https://github.com/leaningtech/labs">
		<img

> 			src="https://img.shields.io/badge/repository-labs-yellow?logo=github"

Reviewer: We should not send people to labs, but to cheerpx-meta

			alt="GitHub repository"
		/>
	</a>
	<a href="https://github.com/leaningtech/labs/issues">
		<img
			src="https://img.shields.io/github/issues/leaningtech/labs?logo=github"
			alt="GitHub issues"
		/>
	</a>
</div>

**CheerpX is an x86 virtualization technology for running executables and operating systems entirely client-side.** It's an always available, zero-cost virtual machine with guaranteed data privacy, powered by a WebAssembly JIT engine.

<div class="flex items-center gap-2 justify-center">
	<LinkButton
		label="WebVM Demo"
		iconLeft="mi:play"
		type="secondary"
		href="https://webvm.io"
	/>
</div>

<div class="m-4 flex justify-center">
	[Get started](/docs/getting-started) · [Guides](/docs/guides) · [API
	Reference](/docs/reference) · [Tutorials](/docs/tutorials) · [Github
	repository](https://github.com/leaningtech/labs)
</div>

```html
<script src="https://cxrtnc.leaningtech.com/1.0.0/cx.js"></script>
```

<div class="text-stone-400">

> 	Not at the right address? Visit our [website](https://leaningtech.com/) to

Reviewer: I am confused by this message. It suggests to me that the loader URL is not the right address. This message (rephrased, something like "Not what you were looking for?" should go above after the very first introduction paragraph

- Without source-code
- Fully client-side

It can also be used to provide a secure sandbox for untrusted code. CheerpX is very robust and deployed as part of [WebVM](https://webvm.io/) and [PythonFiddle](https://pythonfiddle.leaningtech.com/). It is also used in production environments as part of [CheerpX for Flash](https://labs.leaningtech.com/docs/cheerpx-for-flash), an enterprise solution to run legacy Adobe Flash content in the browser.

<div class="text-stone-400">
	If you're looking to run Java in the browser, please check out our specialized
	solution: [CheerpJ](https://cheerpj.com/).
</div>

<LinkButton label="Get started" type="primary" href="/docs/getting-started" />

## Licensing

CheerpX is free to use for personal projects, most non-commercial applications, and for technical evaluations. See [licensing](/docs/licensing) for details.

## How does CheerpX work?

> At its core, CheerpX is a two-tier emulator for the x86 architecture [implemented in C++ and compiled to JavaScript and WebAssembly](https://cheerp.io/). There is both an interpreter and a sophisticated JIT compiler that is able to generate efficient WebAssembly representations for hot code. The whole process is 100% automated and requires no metadata or preprocessing steps. The system is so robust that it can deal with applications that generate new code internally at runtime, and even with self-modifying code.

Reviewer: ```suggestion
At its core, CheerpX is a two-tier emulator for the x86 architecture implemented in C++ and [compiled to JavaScript and WebAssembly](https://cheerp.io/). There is both an interpreter and a sophisticated JIT compiler that is able to generate efficient WebAssembly representations for hot code. The whole process is 100% automated and requires no metadata or preprocessing steps. The system is so robust that it can deal with applications that generate new code internally at runtime, and even with self-modifying code.
```
Link is too long

> This means that you can actually run things like Node.js, including the full V8 engine with its remarkable JIT and code garbage collection.

Reviewer: ```suggestion
This means that you can actually run things such as Node.js, which includes the V8 engine and its remarkable JIT and code garbage collection.
```

> CheerpX currently focuses on user mode (ring 3), and does not fully emulate the kernel (ring 0). We do, however, implement an extended subset of the Linux system call interface, which is sufficient to run most applications. We are also working on a full system emulation mode, with DOS-era applications and games already supported.

Reviewer: ```suggestion
CheerpX currently focuses on user mode code, on top of a Linux-compatible system call interface. Although this layer is not 100% complete it covers an extended subset of Linux capabilities and it is sufficient to run most applications. We are also working on a full system emulation mode, with DOS-era applications and games already supported.
```

Moreover, v86 and JSLinux are hobby projects, whereas CheerpX is a commercial product with a dedicated team of engineers. If you're shipping with CheerpX, you can [rely on us to provide support and maintenance](https://leaningtech.com/contact-us).

## Community

> Have a question, want some inspiration, or just want to chat?

Reviewer: ```suggestion
Have a question, looking for inspiration, or just want to chat?
```
